src/main/java/com/urbanairship/datacube/idservices/HBaseIdService.java

```diff
@@ -4,221 +4,181 @@
-                    }
-                    return id; 
-                case ALLOCATING:
-                    long allocTimestamp = result.getColumn(cf, QUALIFIER).get(0).getTimestamp();
-                    long msSinceAlloc = System.currentTimeMillis() - allocTimestamp;
-                    
-                    if(msSinceAlloc < ALLOC_TIMEOUT_MS) {
-                        // Another thread is already allocating an id for this tag. Wait.
-                        if(log.isDebugEnabled()) {
-                            log.debug("Waiting for other thread to finish allocating id");
-                        }
-                        Thread.sleep(500);
-                        continue;
-                    } else {
-                        log.warn("Preempting expired allocator for input " + 
-                                Base64.encodeBase64String(input));
-                    }
-                    break;
-                default:
-                    throw new RuntimeException("Unexpected column value " + 
-                            Arrays.toString(columnVal));
-                }
-            }
-            // Either (1) there is no entry for this tag, or (2) there was an ALLOCATING
-            // placeholder for this tag, but it has expired. checkAndPut the new value
-            // into place.
-
-            final Put put = new Put(lookupKey);
-            byte[] nanoStamp = Util.longToBytes(System.nanoTime());
-            allocRecord = ArrayUtils.addAll(ALLOCATING_BYTES, nanoStamp);
-            put.add(cf, QUALIFIER, allocRecord);
-            
-            boolean swapSuccess = WithHTable.checkAndPut(pool, lookupTable, lookupKey, cf, 
-                    QUALIFIER, columnVal, put);
-            if(swapSuccess) {
-                if(log.isDebugEnabled()) {
-                    log.debug("Allocation record CAS success");
-                }
-                break;
-            }
-            if(log.isDebugEnabled()) {
-                log.debug("Allocation record CAS failed, retrying");
-            }
-
+        final Optional<byte[]> existingId = getId(lookupKey, numIdBytes);
+        if (existingId.isPresent()) {
+            return existingId.get();
         }
 
-        /*
-         * PHASE 2
-         * 
-         * We've successfully inserted a record showing that we're going to allocate an ID for
-         * this input. This should temporarily exclude other threads from trying to do the
-         * same. This exclusion prevents unused IDs from being allocated if multiple threads
-         * allocate an ID concurrently.
-         * 
-         * Now we assign an ID by taking the next key of the counter for this dimension.
-         */
+        // If not, we need to create it. First get a new id # from the counter table...
         byte[] counterKey = makeCounterKey(dimensionNum);
         final long id = WithHTable.increment(pool, counterTable, counterKey, cf, QUALIFIER, 1L);
```

Comment:
To check my own understanding, I'd like to write down how this occurs:

1. if you attempt to get the id and it exists, sweet, return it.
2. If there is no id, then create it (which is an increment on a cell in hbase)
3. Race to write the id you just got into the counting table. (I didn't succeed in tracing this through to the bit where it compares against nothing, but I think I heard you mention that)
4. If you succeed the race, excellent! return the id you made... after excluding some bytes?
5. If you fail the race, no problem some other thread made it, see what they made, and return it. If they didn't make it, welp, that's real weird.